### PR TITLE
InstanceOf Part Issue #2298 - JEP 455: Primitive Types in Patterns, instanceof, and Switch patterns

### PR DESCRIPTION
**Pull request by mpalat**
This is the initial implementation - this is by no means complete - started and building up

## What it does
 JEP 455 Primitive Types  : Framework and primitive (non-boxing) part  of InstanceOf Testing Contexts issue #2298

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by mpalat** [comment]
This PR addresses the following -
In https://cr.openjdk.org/~abimpoudis/instanceof/jep455-20240424/specs/instanceof-jls.html#jls-5.7

- an identity conversion ([5.1.1](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.1))
- a widening primitive conversion ([5.1.2](https://cr.openjdk.org/~abimpoudis/instanceof/jep455-20240424/specs/instanceof-jls.html#jls-5.1.2))
- a narrowing primitive conversion ([5.1.3](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.3))\a widening and narrowing primitive conversion ([5.1.4](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.4))

Remaining will be addressed in subsequent PRs

**Comment by stephan-herrmann** [comment]
Glancing at this vis-a-vis PR Draft Spec https://cr.openjdk.org/~iris/se/23/spec/pr/java-se-23-pr-spec-01/specs/instanceof-jls.html

@mpalat  could you please roughly correlate your code changes to JLS sections?

In the JLS sections you cite in the above comment I actually don't see any normative changes (disregarding the new terminology "exact widening primitive conversion"  plus informal explanations in [5.1.3](https://cr.openjdk.org/~iris/se/23/spec/pr/java-se-23-pr-spec-01/specs/instanceof-jls.html#jls-5.1.3)).

Thus it's difficult to see, which status has been reached by this PR.

Please don't hesitate to ask for help interpreting the spec! I haven't yet worked through this particular JEP, but if there is any doubt, we should seek clarification very soon!

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 1]
```diff
@@ -207,6 +216,82 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
...
+	this.expression.computeConversion(scope, left, right);
+	int i = this.expression.implicitConversion;
+	codeStream.generateImplicitConversion(i);
+}
+
+private void storeExpressionValue(CodeStream codeStream) {
+	LocalVariableBinding local = this.expression.localVariableBinding();
+	local = local != null ? local : this.secretExpressionValue;
+	if (local != null)
+		codeStream.store(local, this.inPreConstructorContext);
```
[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:292]: In #2685 I'm reading this occurrence of  `inPreConstructorContext` as a poetical approach of saying `false`. Please comment, if you think that some situations will require `true`!

**Comment by mpalat** [comment]
> @mpalat could you please roughly correlate your code changes to JLS sections?

From Section  5.7 Testing Contexts, there are distinctly two logical parts - 
1) If the expression is of a primitive type, then a testing context allows the use of an identity conversion ([5.1.1] ....
2)If the expression is of a reference type, then a testing context allows the use of one of the following:...

This PR addressed the first part

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpalat) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:292]: > In #2685 I'm reading this occurrence of `inPreConstructorContext` as a poetical approach of saying `false`. Please comment, if you think that some situations will require `true`!

I see that you have already made it false.. cool.

**Pull request merged**